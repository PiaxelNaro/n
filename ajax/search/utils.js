// @flow
const mapResultDataObject = (res: { data: Array<any>, more_url?: string }) => {
    let apps;
    if (res.data && res.data.length > 0) {
        apps = res.data.map(app => ({
            id: app.product_id,
            icon: app.icon,
            company: app.publisher_name,
            name: app.product_name,
            link: app.target_url,
            store: app.market,
            flagCode: app.country_code,
            showStoreIcon: true,
        }));
        return {
            results: apps,
            moreUrl: res.more_url || '',
        };
    }
    return {
        results: [],
    };
};

export default mapResultDataObject;
