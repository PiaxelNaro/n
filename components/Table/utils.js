// @flow

export type Align = 'left' | 'center' | 'right';

export type LooseColumn = {
    key?: string,
    keys?: Array<string>,
    label?: any,
    tooltip?: any,
    subColumns?: Array<LooseColumn>,
    size?: Array<string> | string,
    align?: Array<Align> | Align,
    isSortable?: boolean,
};

export type Column = {
    key: string,
    keys: Array<string>,
    label: any,
    tooltip: any,
    subColumns: Array<Column>,
    size: Array<string>,
    isSortable: boolean,
    align: Array<Align>,
    headerAlign: Align,
};

export const getColumnColSpan = (col: Column) =>
    Math.max(
        1,
        col.keys.length,
        col.subColumns.reduce((m, { keys }) => m + keys.length, 0)
    );

export const getColumnRowSpan = (col: Column) => (col.subColumns.length ? 1 : 2);

export const normalizeColumn = (column: LooseColumn): Column => {
    const subColumns = column.subColumns ? column.subColumns.map(normalizeColumn) : [];
    let keys = [];
    if (!subColumns.length) {
        keys = column.keys || [];
        if (!keys.length) {
            keys = column.key ? [column.key] : [];
        }
    }
    const size = keys.map((_, i) => {
        if (Array.isArray(column.size)) {
            return i in column.size ? column.size[i] : 'medium';
        }

        return column.size ? column.size : 'medium';
    });

    const align = keys.map((_, i) => {
        if (Array.isArray(column.align)) {
            return i in column.align ? column.align[i] : 'left';
        }

        return column.align ? column.align : 'left';
    });

    let headerAlign = 'left';
    if (subColumns.length <= 1 && align.length === 1) {
        // follow column align if not composed column
        headerAlign = align[0];
    }

    let key = column.key || keys.join('|');
    if (subColumns.length) {
        // eslint-disable-next-line no-shadow
        key = subColumns.map(({ key }) => key).join('|');
    }

    return {
        key,
        isSortable: Boolean(column.isSortable),
        label: column.label,
        tooltip: column.tooltip,
        subColumns,
        keys,
        size,
        headerAlign,
        align,
    };
};

function getFirstHTMLChildren(tr: HTMLElement): void | HTMLElement {
    return Array.from(tr.children).find(node => Boolean(node.tagName));
}

export const createSizeMatrix = (tableScroller: ?HTMLElement) => {
    const table = tableScroller ? tableScroller.querySelector('table') : null;

    if (!tableScroller || !table) {
        return {
            width: 'auto',
            scrollerWidth: 'auto',
            rows: [],
        };
    }

    const mainTrs = Array.from(table.querySelectorAll('tr'));
    return {
        width: `${table.offsetWidth}px`,
        scrollerWidth: `${tableScroller.getBoundingClientRect().width}px`,
        rows: mainTrs.map((tr: HTMLElement) => {
            let cells = [];
            if (tr.parentNode && tr.parentNode.tagName === 'THEAD') {
                cells = Array.from(tr.children).map(cell => ({
                    width: `${cell.offsetWidth}px`,
                }));
            }

            // Chrome mess up rowspan heights, so we need to rely on the first child size if the
            // row has no height or no first child.
            // Here we won't get height from element's offsetHeight
            // because table cell size in Firefox will contain decimal number
            let rowHeight = tr.getBoundingClientRect().height;
            const firstChild = getFirstHTMLChildren(tr);
            if (
                rowHeight === 0 &&
                firstChild &&
                (firstChild.tagName === 'TH' || firstChild.tagName === 'TD')
            ) {
                rowHeight = firstChild.getBoundingClientRect().height;
            } else if (rowHeight !== 0 && !firstChild) {
                rowHeight = 0;
            }
            return {
                height: `${rowHeight}px`,
                cells,
            };
        }),
    };
};
