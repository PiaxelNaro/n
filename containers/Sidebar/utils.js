// @flow
import { createSelector } from 'reselect';
import { type I18nType } from '../../modules/i18n';
import { type UserType } from '../../modules/user';

// If user is Non-ANA, empty sub navs
const getANASidebar = (user: UserType, i18n: I18nType) => {
    if (user.isANA) {
        return [
            {
                id: 'c',
                title: '',
                items: [
                    {
                        id: 'c1',
                        title: i18n.gettext('帮助'),
                        href: '/help/',
                        isVisible: true,
                    },
                    {
                        id: 'c2',
                        title: i18n.gettext('课程'),
                        href: '/course',
                        isVisible: true,
                    },
                ],
            },
        ];
    }
    return null;
};

// Store PP locked when
// 1. user does not buy any store contract
// 2. user only buy store PYA or store PYP
const isStorePPLocked = (user: UserType) =>
    !user.isINT || !user.intAuth.store || !user.intAuth.store.pp;

// Store PYA invisible when
// 1. user does not buy any store contract
// 2. user is store PP only or store PYP only
const isStorePYAUser = (user: UserType) =>
    user.isINT && user.intAuth.store && user.intAuth.store.pya;

// Store PYP visible when
// user is store PYA or store PYP
const isStorePYPVisible = (user: UserType) =>
    user.isINT &&
    user.intAuth.store &&
    (user.intAuth.store.pyp || user.intAuth.store.pya);

// Store PYP locked when
// 1. user does not buy store contract
// 2. user is not store PYP
const isStorePYPLocked = (user: UserType) =>
    !user.isINT || !user.intAuth.store || !user.intAuth.store.pyp;

// Usage PP unlocked when user buy usage pp
const isUsagePPLocked = (user: UserType) =>
    !user.isINT || !user.intAuth.usage || !user.intAuth.usage.pp;

// Usage PYA show and unlocked if user buy usage pya
const isUsagePYAUser = (user: UserType) =>
    user.isINT && user.intAuth.usage && user.intAuth.usage.pya;

// Usage retention unlocked if user buy usage pp
const isUsagePPUser = (user: UserType) =>
    user.isINT && user.intAuth.usage && user.intAuth.usage.pp;

const isMarketAsoUser = (user: UserType) =>
    user.isINT && user.intAuth.marketing && user.intAuth.marketing.aso;

const isMarketAdvertisingUser = (user: UserType) =>
    user.isINT && user.intAuth.marketing && user.intAuth.marketing.advertising;

const getINTSidebar = (user: UserType, i18n: I18nType) =>
    [
        {
            id: 'a',
            title: i18n.gettext('排课'),
            items: [
                {
                    id: 'a1',
                    title: i18n.gettext('选课'),
                    href: '/#',
                    isVisible: true,
                    isLocked: !user.isINT,
                },
            ],
        },
        {
            id: 'b',
            title: i18n.gettext('查询'),
            items: [
                {
                    id: 'b1',
                    title: i18n.gettext('课程表'),
                    href: '/apps/ios/top-chart/',
                    isVisible: true,
                },
            ],
        }
    ].map(child => ({
        ...child,
        items: child.items.filter(item => item.isVisible),
    }));

const getSidebar = createSelector(
    user => user,
    (user, i18n) => i18n,
    (user: UserType, i18n: I18nType) => {
        const ANA_NAV = getANASidebar(user, i18n);
        const INT_NAV = getINTSidebar(user, i18n);
        return {
            current: 'a1',
            home: {
                id: 'home',
                title: i18n.gettext('首页'),
                href: '/home',
                isVisible: true,
            },
            platforms: [
                {
                    id: 'my_app_analytics',
                    title: i18n.gettext('学生功能'),
                    href: '/f',
                    icon: 'aa-connections',
                    items: ANA_NAV,
                },
                {
                    id: 'market_intelligence',
                    title: i18n.gettext('管理员功能'),
                    items: INT_NAV,
                },
            ],
        };
    }
);

export default getSidebar;
