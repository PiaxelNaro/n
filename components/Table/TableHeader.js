// @flow
/* eslint react/no-array-index-key: 0 */

import React from 'react';
import styled from 'styled-components';
import noop from 'lodash/noop';
import classnames from 'classnames';
import FlexGrid from '../FlexGrid';
import SortIndicator from '../SortIndicator';
import { type Column, getColumnColSpan, getColumnRowSpan } from './utils';
import Tooltip from '../Overlay/Tooltip';

export type SortingRule = { key?: string, dir?: 'asc' | 'desc' | null };

const borderStyle = '1px solid #D9DCDF';
const Th = styled.th`
    border: ${borderStyle};

    &:first-child {
        border-left: none;
    }

    &:last-child {
        border-right: none;
    }

    &.isSortable {
        cursor: pointer;
    }

    &.isSortable:hover,
    &.isSortable:focus {
        background-color: #f9f9f9;
    }
`;

const TableHead = styled.thead`
    background: #efefef;
    font-weight: bold;
    color: #000;
`;

const HeaderCell = ({
    children,
    onSort,
    sortDirection,
    colSpan = 1,
    rowSpan = 1,
    style,
    align = 'left',
}: {
    children?: any,
    onSort?: ?() => any,
    sortDirection?: 'asc' | 'desc' | null,
    colSpan?: number,
    rowSpan?: number,
    style?: any,
    align?: 'left' | 'center' | 'right',
}) => {
    const isSortable = Boolean(onSort);

    let sortIcon;
    if (isSortable) {
        sortIcon = (
            <FlexGrid.Item flex={0}>
                <SortIndicator direction={sortDirection} />
            </FlexGrid.Item>
        );
    }
    return (
        <Th
            onClick={onSort}
            colSpan={colSpan}
            rowSpan={rowSpan}
            style={style}
            className={classnames({ isSortable })}
        >
            <FlexGrid gutter="4px">
                <FlexGrid.Item
                    style={{
                        textAlign: align,
                    }}
                    flex={1}
                >
                    {children}
                </FlexGrid.Item>
                {sortIcon}
            </FlexGrid>
        </Th>
    );
};

const HeaderCellContent = ({ label, tooltip }: { label: any, tooltip?: any }) => {
    if (!tooltip) {
        return <span>{label}</span>;
    }
    return <Tooltip content={tooltip}>{label}</Tooltip>;
};

const TableHeader = ({
    columns,
    sizeMatrix,
    onSort = noop,
    sortedBy = {},
}: {
    columns: Array<Column>,
    sizeMatrix?: Object,
    onSort?: (string, string) => any,
    sortedBy?: SortingRule,
}) => {
    const getRowStyle = rowIndex => {
        if (sizeMatrix) {
            return { height: sizeMatrix.rows[rowIndex].height };
        }

        return {};
    };

    const getCellStyle = (rowIndex, cellIndex) => {
        if (sizeMatrix) {
            return { width: sizeMatrix.rows[rowIndex].cells[cellIndex].width };
        }

        return {};
    };

    const handleClick = key => {
        if (key !== sortedBy.key) {
            onSort(key, 'asc');
        } else {
            onSort(key, sortedBy.dir === 'asc' ? 'desc' : 'asc');
        }
    };

    const subColumns = columns.reduce((m, col) => m.concat(col.subColumns), []);

    return (
        <TableHead>
            {[columns, subColumns].map((cells, rowIndex) =>
                <tr style={getRowStyle(rowIndex)} key={rowIndex}>
                    {cells.map((col, cellIndex) =>
                        <HeaderCell
                            key={col.key}
                            align={col.headerAlign}
                            colSpan={getColumnColSpan(col)}
                            rowSpan={getColumnRowSpan(col)}
                            style={getCellStyle(rowIndex, cellIndex)}
                            onSort={
                                col.isSortable ? handleClick.bind(null, col.key) : null
                            }
                            sortDirection={col.key === sortedBy.key ? sortedBy.dir : null}
                        >
                            <HeaderCellContent label={col.label} tooltip={col.tooltip} />
                        </HeaderCell>
                    )}
                </tr>
            )}
        </TableHead>
    );
};

export default TableHeader;
export { HeaderCell, HeaderCellContent };
