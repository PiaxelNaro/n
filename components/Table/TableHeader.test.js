// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { normalizeColumn } from './utils';
import TableHeader, { HeaderCell, HeaderCellContent } from './TableHeader';

describe('<HeaderCell>', () => {
    it('can be sortable', () => {
        const spy = jest.fn();
        const cell = shallow(
            <HeaderCell onSort={spy} sortDirection="desc">
                Content
            </HeaderCell>
        );

        expect(cell).toMatchSnapshot();

        cell.simulate('click');
        expect(spy).toHaveBeenCalledTimes(1);
    });

    it('can handle layout properties', () => {
        expect(
            shallow(
                <HeaderCell colSpan={2} rowSpan={2} style={{ width: '30px' }}>
                    Content
                </HeaderCell>
            )
        ).toMatchSnapshot();
    });
});

describe('<HeaderCellContent />', () => {
    it('renders label', () => {
        expect(shallow(<HeaderCellContent label="Lorem ipsum." />)).toMatchSnapshot();
    });

    it('renders label and tooltip', () => {
        expect(
            shallow(
                <HeaderCellContent
                    label="Lorem ipsum."
                    tooltip="Lorem ipsum dolor sit."
                />
            )
        ).toMatchSnapshot();
    });
});

describe('<TableHeader />', () => {
    const columns = [
        { key: 'bar', label: 'Bar' },
        {
            label: 'Foo',
            subColumns: [
                { key: 'foo1', label: 'Foo 1' },
                { key: 'foo2', label: 'Foo 2' },
            ],
        },
    ].map(normalizeColumn);

    it('renders multi level table header', () => {
        const header = shallow(<TableHeader columns={columns} />);
        expect(header).toMatchSnapshot();
    });

    it('renders fixed size header', () => {
        const matrix = {
            rows: [
                {
                    height: '1px',
                    cells: [{ width: '2px' }, { width: '3px' }],
                },
                {
                    height: '4px',
                    cells: [{ width: '5px' }, { width: '6px' }],
                },
            ],
        };

        const header = shallow(<TableHeader columns={columns} sizeMatrix={matrix} />);
        expect(header).toMatchSnapshot();
    });

    it('handles sorting', () => {
        const sortableColumns = [
            { key: 'bar', label: 'Bar', isSortable: true },
            {
                label: 'Foo',
                subColumns: [
                    { key: 'foo1', label: 'Foo 1', isSortable: true },
                    { key: 'foo2', label: 'Foo 2', isSortable: true },
                ],
            },
        ].map(normalizeColumn);

        const spy = jest.fn();
        let header = shallow(
            <TableHeader
                columns={sortableColumns}
                onSort={spy}
                sortedBy={{ key: 'bar', dir: 'asc' }}
            />
        );

        const cells = header.find('HeaderCell');

        cells.at(0).prop('onSort')();
        expect(spy.mock.calls[0]).toEqual(['bar', 'desc']);

        cells.at(2).prop('onSort')();
        expect(spy.mock.calls[1]).toEqual(['foo1', 'asc']);

        // Test with desc direction:
        header = shallow(
            <TableHeader
                columns={sortableColumns}
                onSort={spy}
                sortedBy={{ key: 'bar', dir: 'desc' }}
            />
        );
        header.find('HeaderCell').at(0).prop('onSort')();
        expect(spy.mock.calls[2]).toEqual(['bar', 'asc']);
    });
});
