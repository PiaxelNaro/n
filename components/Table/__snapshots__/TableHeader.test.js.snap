// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderCell> can be sortable 1`] = `
<TableHeader__Th
  className="isSortable"
  colSpan={1}
  onClick={[Function]}
  rowSpan={1}
>
  <FlexGrid
    gutter="4px"
  >
    <FlexGrid.Item
      flex={1}
      style={
        Object {
          "textAlign": "left",
        }
      }
    >
      Content
    </FlexGrid.Item>
    <FlexGrid.Item
      flex={0}
    >
      <SortIndicator
        direction="desc"
      />
    </FlexGrid.Item>
  </FlexGrid>
</TableHeader__Th>
`;

exports[`<HeaderCell> can handle layout properties 1`] = `
<TableHeader__Th
  className=""
  colSpan={2}
  rowSpan={2}
  style={
    Object {
      "width": "30px",
    }
  }
>
  <FlexGrid
    gutter="4px"
  >
    <FlexGrid.Item
      flex={1}
      style={
        Object {
          "textAlign": "left",
        }
      }
    >
      Content
    </FlexGrid.Item>
  </FlexGrid>
</TableHeader__Th>
`;

exports[`<HeaderCellContent /> renders label 1`] = `
<span>
  Lorem ipsum.
</span>
`;

exports[`<HeaderCellContent /> renders label and tooltip 1`] = `
<Tooltip
  content="Lorem ipsum dolor sit."
>
  Lorem ipsum.
</Tooltip>
`;

exports[`<TableHeader /> renders fixed size header 1`] = `
<TableHeader__TableHead>
  <tr
    style={
      Object {
        "height": "1px",
      }
    }
  >
    <HeaderCell
      align="left"
      colSpan={1}
      onSort={null}
      rowSpan={2}
      sortDirection={null}
      style={
        Object {
          "width": "2px",
        }
      }
    >
      <HeaderCellContent
        label="Bar"
      />
    </HeaderCell>
    <HeaderCell
      align="left"
      colSpan={2}
      onSort={null}
      rowSpan={1}
      sortDirection={null}
      style={
        Object {
          "width": "3px",
        }
      }
    >
      <HeaderCellContent
        label="Foo"
      />
    </HeaderCell>
  </tr>
  <tr
    style={
      Object {
        "height": "4px",
      }
    }
  >
    <HeaderCell
      align="left"
      colSpan={1}
      onSort={null}
      rowSpan={2}
      sortDirection={null}
      style={
        Object {
          "width": "5px",
        }
      }
    >
      <HeaderCellContent
        label="Foo 1"
      />
    </HeaderCell>
    <HeaderCell
      align="left"
      colSpan={1}
      onSort={null}
      rowSpan={2}
      sortDirection={null}
      style={
        Object {
          "width": "6px",
        }
      }
    >
      <HeaderCellContent
        label="Foo 2"
      />
    </HeaderCell>
  </tr>
</TableHeader__TableHead>
`;

exports[`<TableHeader /> renders multi level table header 1`] = `
<TableHeader__TableHead>
  <tr
    style={Object {}}
  >
    <HeaderCell
      align="left"
      colSpan={1}
      onSort={null}
      rowSpan={2}
      sortDirection={null}
      style={Object {}}
    >
      <HeaderCellContent
        label="Bar"
      />
    </HeaderCell>
    <HeaderCell
      align="left"
      colSpan={2}
      onSort={null}
      rowSpan={1}
      sortDirection={null}
      style={Object {}}
    >
      <HeaderCellContent
        label="Foo"
      />
    </HeaderCell>
  </tr>
  <tr
    style={Object {}}
  >
    <HeaderCell
      align="left"
      colSpan={1}
      onSort={null}
      rowSpan={2}
      sortDirection={null}
      style={Object {}}
    >
      <HeaderCellContent
        label="Foo 1"
      />
    </HeaderCell>
    <HeaderCell
      align="left"
      colSpan={1}
      onSort={null}
      rowSpan={2}
      sortDirection={null}
      style={Object {}}
    >
      <HeaderCellContent
        label="Foo 2"
      />
    </HeaderCell>
  </tr>
</TableHeader__TableHead>
`;
